pkg/bearertoken/grpc.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package bearertoken
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const Key = "bearer.token"
+
+type tokenatedServerStream struct {
+	grpc.ServerStream
+	context context.Context
+}
+
+func (tss *tokenatedServerStream) Context() context.Context {
+	return tss.context
+}
+
+// getValidBearerToken attempts to retrieve the bearer token from the context.
+// It does not return an error if the token is missing.
+func getValidBearerToken(ctx context.Context, bearerHeader string) (string, error) {
+	bearerToken, ok := GetBearerToken(ctx)
+	if ok && bearerToken != "" {
+		return bearerToken, nil
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", nil
+	}
+
+	tokens := md.Get(bearerHeader)
+	if len(tokens) < 1 {
+		return "", nil
+	}
+	if len(tokens) > 1 {
+		return "", fmt.Errorf("malformed token: multiple tokens found")
+	}
+	return tokens[0], nil
+}
+
+// NewStreamServerInterceptor creates a new stream interceptor that injects the bearer token into the context if available.
+func NewStreamServerInterceptor() grpc.StreamServerInterceptor {
+	return streamServerInterceptor
```

Comment:
please use the same pattern in all four scenarios - either a separate function or a nested function.